### PR TITLE
set the graphics context width and height by the Vizkit3DWidget constructor

### PR DESCRIPTION
**Pull request by gustavoneves12**
To change the simulated camera size is necessary change the graphics context size.

### CODE REVIEW

**Comment by doudou** [comment]
I would prefer that we merge this directly on the rock-core repository, and the re-merge the default manipulator branch on our fork ... It's just to avoid adding more private stuff to our private stuff.

**Comment by doudou** [comment]
Ping ... what's the status on that ?

**Comment by gustavoneves12** [comment]
This is working. 
Currently we can change the display size.  But to achieve this, it is necessary set the graphics context size. We cannot change the graphics context size after its creation.

**Comment by doudou** [comment]
> This is working. 

Not the "that" I was referring to. I meant

> I would prefer that we merge this directly on the rock-core repository, and the re-merge the default 
> manipulator branch on our fork ... It's just to avoid adding more private stuff to our private stuff.

**Comment by gustavoneves12** [comment]
I am going to close this PR and re-open this PR to rock-core repository. Ok?

**Comment by doudou** [comment]
Yes. You will have to make sure that you don't add the "default manipulator" that is not ready to go there ...

**Comment by doudou** [comment]
(i.e. just make sure to add only your commit)

**Comment by gustavoneves12** [comment]
@doudou there are some conflicts between this fork and the rock-core repository. I changed the remote to rock-core repository. Afterward the simulated cameras stopped to work. I am going to look into this and try to detect the problem with the cameras. 

@doudou what is the vizkit3d repository we must use?

**Comment by doudou** [comment]
Actually, Alex merged all our changes back to the rock-core repository. We're virtually in the same state. We should stop using our fork and use the rock-core repository instead.

**Comment by gustavoneves12** [comment]
There is something strange in the rock-core repository. 
The output image is inverted and the color channels are wrong as well.

**Camera vizkit3d bir-fork**
camera_vizkit3d_fork

**Camera vizkit3d rock-core**
camera_vizkit3d_rock-core

**Comment by gustavoneves12** [comment]
The commit below are not in the rock-core repository:

https://github.com/Brazilian-Institute-of-Robotics/gui-vizkit3d/commit/5204ed5e97f6e5e00feecd068a5532384f5e29ca

https://github.com/Brazilian-Institute-of-Robotics/gui-vizkit3d/commit/c3b3abb508610e887770123c3a5be1aa5ea78884

**Comment by doudou** [comment]
Merge commits are not a problem. The issue is that the image grabbing PR has been merged but then reverted because it's not compatible with OSG 3.4. We really depend on these image grabbing patches.

So, the bottom line is that we need to reset hard on rock-core master and then just cherry-pick the image grabbing commit. https://github.com/Brazilian-Institute-of-Robotics/gui-vizkit3d/tree/sync_with_rock_core

Test https://github.com/Brazilian-Institute-of-Robotics/gui-vizkit3d/tree/sync_with_rock_core, and if it works fine reset master to it and force push to our fork.

**Comment by gustavoneves12** [comment]
Ok @doudou

**Comment by doudou** [comment]
(Important note: the branch I just made does NOT contain your change, you have to merge the PR branch manually as well)

**Comment by gustavoneves12** [comment]
I already tested.
I am going to merge it.

**Comment by doudou** [comment]
> I am going to merge it.

Just to be sure: _replace_ our current master by this branch and then _merge_ the width/height change, OK ?

**Comment by gustavoneves12** [comment]
I merged everything in the master. I am testing now.

**Comment by gustavoneves12** [comment]
There are some pending PR: https://github.com/Brazilian-Institute-of-Robotics/simulation-orogen-underwater_camera_simulation/pull/7.

If I merged the width/height change,  before we merge that pendings PR, we will get the error below:

``` bash
widget = new vizkit3d::Vizkit3DWidget(NULL, "world_osg", false);
                                                                       ^
    In file included from /home/gustavoneves/flat_fish/dev/gui/vizkit3d_world/build/include/_vizkit3d_world_/vizkit3d_world/Vizkit3dWorld.hpp:14:0,
                     from /home/gustavoneves/flat_fish/dev/gui/vizkit3d_world/src/Vizkit3dWorld.cpp:14:
    /home/gustavoneves/flat_fish/dev/install/include/vizkit3d/Vizkit3DWidget.hpp:190:13: error:   initializing argument 2 of ‘vizkit3d::Vizkit3DWidget::Vizkit3DWidget(QWidget*, int, int, const QString&, bool)’ [-fpermissive]
                 Vizkit3DWidget(QWidget* parent = 0, int width = 800, int height = 600,
                 ^
    make[2]: *** [src/CMakeFiles/vizkit3d_world.dir/Vizkit3dWorld.cpp.o] Error 1
```

**Comment by doudou** [comment]
Yes, we need to merge all of them together.

**Comment by doudou** [comment]
And so ping Diego

**Comment by doudou** [comment]
As soon as you're ready, push the updated branch to master and merge all the PRs
- https://github.com/Brazilian-Institute-of-Robotics/gui-vizkit3d_world/pull/8
- https://github.com/Brazilian-Institute-of-Robotics/gui-orogen-vizkit3d_world/pull/5
- https://github.com/Brazilian-Institute-of-Robotics/simulation-orogen-underwater_camera_simulation/pull/7

What about the sonar simulation ? Would it need to use that parameter as well ? @romulogcerqueira

**Comment by romulogcerqueira** [comment]
@doudou , no problems for sonar simulation. I set the width/height using **setCameraParams()** method.

**Comment by doudou** [comment]
Actually, as far as I understood, setCameraParams only scales the viewport image onto whatever target you have, but does not change the actual rendering viewport size. I.e. you're not controlling the most important parameter ... (and might be doing either too much or too little work for the sonar simulation's goal. In the case of the micron, I can safely say "way too much")

**Comment by romulogcerqueira** [comment]
My mistake, @doudou . I was talking with @trocolit2 about this process.

I have used the **ImageViewerCaptureTool** class to define the viewport as follows:

`  /**
     \* @brief This class generate a hide viewer to get the osg::image without GUI.
     *
     \*  @param width: Width to generate the image
     \*  @param height: height to generate the image
     */
    ImageViewerCaptureTool(uint width = 640, uint height = 480);

```
/**
 * @brief This constructor class generate a image according fovy, fovx and height resolution.
 *
 *  @param fovy: vertical field of view
 *  @param fovx: horizontal field of view
 *  @param height: height to generate the image
 */

ImageViewerCaptureTool(double fovY, double fovX, uint value, bool isHeight = true);`
```

The `setCameraParams()` is only used to change the camera parameters (i.e. focus and FOVs).

**Comment by doudou** [comment]
The problem, as far as I understood the point of these patches, is that the image capture only scales the viewport to the with/height specification. It does not change how many pixels are actually rendered (please correct me if I am wrong @gustavoneves12 and @dbcesar)

In other words, if you set up the viewport with 1024x1024, your single-beam micron will have to both render and process 1M pixels, regardless of how you call ImageViewerCaptureTool. On the other hand,  if you set up the viewport with the same 1024x1024 for the gemini, you get 8 pixels on the width per beam but 1024 for the 30 degree vertical opening.

My guess is that you would need to somehow find a good "pixel per angle" constant and use it _both_ when setting up vizkit3d_world _and_ during capture. I.e. if you say that you need one pixel per 0.1 degree (0.002 radians, 20cm at 100m) for a good result, the viewport should be 1200x200 for the gemini and 30x350 for the micron. We'll need to tune once we actually do something with the data, but I guess that these values would be decent values for now.

TL;DR: don't render pixels you won't use.
